packages/client/src/object/Cache.ts

```diff
@@ -92,18 +105,23 @@ export function createAsyncCache<K, V extends {}>(
   fn: (client: MinimalClient, key: K) => Promise<V>,
 ): AsyncCache<K, V> {
   const cache = createCache<K, V>();
+  const inProgress = createCache<K, Promise<V> | V>();
 
   function getOrUndefined(client: MinimalClient, key: K) {
     return cache.get(client, key);
   }
 
   async function get(client: MinimalClient, key: K) {
-    return cache.get(client, key)
+    return cache.get(client, key) ?? inProgress.get(client, key)
       ?? set(client, key, fn(client, key));
   }
 
   async function set(client: MinimalClient, k: K, v: V | Promise<V>) {
-    return cache.set(client, k, await v);
```

Comment:
Can you avoid needing `inProgress` by making this something like:

```
cache.set(client,k,v);
return await v;
```

I've had issues trying to make sure this sort of caching doesn't have subtle race conditions, so simpler is better.

Reply:
wrote a ton of tests for you in the upcoming pr

Reply:
I'm relying on the `getOrUndefined` to return instantly. I guess I don't have to but I like the semantics better.